airbyte-db/db-lib/src/main/java/io/airbyte/db/factory/DataSourceFactory.java

```diff
@@ -63,6 +63,32 @@ public static DataSource create(final String username,
         .build();
   }
 
+  /**
+   * Constructs a new {@link DataSource} using the provided configuration and extracts
+   * connectionTimeout (as seconds) if it exists
+   *
+   * @param username The username of the database user.
+   * @param password The password of the database user.
+   * @param driverClassName The fully qualified name of the JDBC driver class.
+   * @param jdbcConnectionString The JDBC connection string.
+   * @param connectionProperties Additional configuration properties for the underlying driver.
+   * @return The configured {@link DataSource}.
+   */
+  public static DataSource createWithConnectionTimeout(final String username,
+                                                       final String password,
+                                                       final String driverClassName,
+                                                       final String jdbcConnectionString,
+                                                       final Map<String, String> connectionProperties) {
```

Comment:
This extra method does not seem necessary. The same logic can be put in the `create` method to extract the connection ms timeout if it exists in `connectionProperties`.

There are two downsides of introducing a new method:
1. People won't know that they need to use this method to construct a `DataSource`.
2. In the future, when we want to process another parameter, another method will be introduced if we follow this pattern, and then we will have a combination problem. Does the new method include both `connectionTimeout` and the new parameter, or should there be two new methods - one only for the new parameter, and one for both?

Reply:
Yeah, I agree. The basis for creating this new method `createWithConnectionTimeout` was primarily due to the single responsibility principle for keeping `withConnectionProperties` only limited to `connectionProperties`. The "best" solution in my mind would be to have `connectionTimeout` as an explicit parameter such that the code utilizes`withConnectionTimeoutMs` for the `DataSource` build.

As a I think about it, this `createDataSource` _should_ be a method that is overloaded in the `Postgres Source` since the reason that `connectionProperties` does not pass in `connectionTimeout` is because Postgres defines `connectTimeout` as a field whereas Hikari defines `connectionTimeout` when the code executes `connectionProperties.forEach(config::addDataSourceProperty)` the mapping isn't correct so the default of 60000 ms remains

Even with overloading this method in Postgres, it seems to make sense in my mind to keep this `createWithconnectionTimeout` for the usages of `Postgres Source` only but if you still believe the logic should be in `withConnectionProperties` then I would still call the `getConnectionTimeoutMs` method within `withConnectionProperties` and wanted to avoid having that method do more than just setting the `connectionProperties`

Does that make sense for why I created a new createDataSource method?

Reply:
> The basis for creating this new method createWithConnectionTimeout was primarily due to the single responsibility principle for keeping `withConnectionProperties` only limited to `connectionProperties`.

Agree that `withConnectionProperties` should have a single responsibility. The responsibility of extracting `connectionTimeout` and setting it in the builder should live inside either the `create` method or the client (e.e. `PostgresSource`).

----

> The "best" solution in my mind would be to have `connectionTimeout` as an explicit parameter such that the code utilizes `withConnectionTimeoutMs` for the DataSource build.

Without a separate method, we can still utilize the `withConnectionTimeoutMs` method in the existing `create` method:

```java
public static DataSource create(final String username,
                                final String password,
                                final String driverClassName,
                                final String jdbcConnectionString,
                                final Map<String, String> connectionProperties) {
  final DataSourceBuilder builder = new DataSourceBuilder()
      .withConnectionProperties(connectionProperties)
      .withDriverClassName(driverClassName)
      .withJdbcUrl(jdbcConnectionString)
      .withPassword(password)
      .withUsername(username)
      .build();
  }

  if (connectionProperties.containsKey(CONNECT_TIMEOUT_KEY)) {
    builder = builder.withConnectionTimeoutMs(Long.parseLong(connectionProperties.get(CONNECT_TIMEOUT_KEY)));
  }

  return builder;
}
```

What do you think about this approach?

----

> this `createDataSource` should be a method that is overloaded in the `PostgresSource`

If we go with this alternative approach, the client, i.e. `PostgresSource`, can extract the `connectTimeout` parameter and explicit call the `withConnectionTimeoutMs` method. Something like:

```java
final DataSource dataSource = DataSourceFactory.create(
    jdbcConfig.has(JdbcUtils.USERNAME_KEY) ? jdbcConfig.get(JdbcUtils.USERNAME_KEY).asText() : null,
    jdbcConfig.has(JdbcUtils.PASSWORD_KEY) ? jdbcConfig.get(JdbcUtils.PASSWORD_KEY).asText() : null,
    driverClass,
    jdbcConfig.get(JdbcUtils.JDBC_URL_KEY).asText(),
    getConnectionProperties(config));

if (connectionProperties.containsKey(CONNECT_TIMEOUT_KEY)) {
  dataSource = dataSource.withConnectionTimeoutMs(Long.parseLong(connectionProperties.get(CONNECT_TIMEOUT_KEY)));
}
dataSources.add(dataSource);
return dataSource;
```

But ideally we should avoid overloading `createDataSource` in `PostgresSource`.

Reply:
I'll trial both of these approaches but I did not think much about adding this logic to the `AbstractJdbcSource#createDataSource` so thanks for sharing this thought process

I'm preemptively leaning towards creating this in `PostgresSource` since the failure is due to the `key`s being named _slightly_ different but if other connections use the same key as Hikari `connectionTimeout` then they would not have issues so the current logic for `AbstractJdbcSource#createDataSource` should not change in my mind

I also just confirmed the `MySQL` and `MSSQL` both use `connectionTimeout` so they will properly override from the `connectionProperties` without any changes to the current code